File: common/src/main/java/net/shiroha233/roadweaver/util/ComputeService.java
package net.shiroha233.roadweaver.util;

import net.shiroha233.roadweaver.runtime.ThreadPoolManager;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.Executor;
import java.util.function.Supplier;

/**
 * 异步计算服务，封装共享工作池的提交操作
 */
public final class ComputeService {

    private ComputeService() {}

    public static Executor executor() {
        return ThreadPoolManager.roleExecutor(ThreadPoolManager.TaskRole.PLANNING);
    }

    public static Executor executor(ThreadPoolManager.TaskRole role) {
        return ThreadPoolManager.roleExecutor(role);
    }

    public static Executor mapExecutor() {
        return ThreadPoolManager.roleExecutor(ThreadPoolManager.TaskRole.MAP);
    }

    public static <T> CompletableFuture<T> supplyAsync(Supplier<T> supplier) {
        return supplyAsync(ThreadPoolManager.TaskRole.PLANNING, supplier);
    }

    public static <T> CompletableFuture<T> supplyAsync(ThreadPoolManager.TaskRole role, Supplier<T> supplier) {
        return ThreadPoolManager.supplyAsync(role, supplier);
    }

    public static CompletableFuture<Void> runAsync(Runnable runnable) {
        return runAsync(ThreadPoolManager.TaskRole.PLANNING, runnable);
    }

    public static CompletableFuture<Void> runAsync(ThreadPoolManager.TaskRole role, Runnable runnable) {
        return ThreadPoolManager.runAsync(role, runnable);
    }

    public static CompletableFuture<Void> runMapAsync(Runnable runnable) {
        return runAsync(ThreadPoolManager.TaskRole.MAP, runnable);
    }

    public static void shutdownNow() {
        ThreadPoolManager.onServerStopping();
    }
}

File: common/src/main/kotlin/net/shiroha233/roadweaver/util/ComputeService.kt
package net.shiroha233.roadweaver.util

import net.shiroha233.roadweaver.runtime.ThreadPoolManager
import java.util.concurrent.CompletableFuture
import java.util.concurrent.Executor
import java.util.function.Supplier

@Suppress("MemberVisibilityCanBePrivate")
object ComputeService {
    @JvmStatic
    fun executor(): Executor {
        return ThreadPoolManager.computeExecutor()
    }

    @JvmStatic
    fun <T> supplyAsync(supplier: Supplier<T>): CompletableFuture<T> {
        return CompletableFuture.supplyAsync(supplier, ThreadPoolManager.computeExecutor())
    }

    @JvmStatic
    fun runAsync(runnable: Runnable): CompletableFuture<Void> {
        return CompletableFuture.runAsync(runnable, ThreadPoolManager.computeExecutor())
    }

    @JvmStatic
    fun shutdownNow() {
        // 委托中央管理器以确保纪元轮换和完整的池关闭
        ThreadPoolManager.onServerStopping()
    }
}
